File: Anul-2/Semestrul I/MAP/Seminar/sem13/src/main/java/org/example/sem13/repository/InFileRepository.java
package org.example.sem13.repository;

import org.example.sem13.domain.Entity;

public abstract class InFileRepository<ID, E extends Entity<ID>> extends InMemoryRepository<ID, E> implements Repository<ID, E> {
    protected String fileName;

    abstract void loadFromFile();

    public InFileRepository(String fileName) {
        super();
        this.fileName = fileName;
    }
}

File: Anul-2/Semestrul I/MAP/Seminar/Seminar-11/Seminar-11/InFileRepository.cs
namespace Seminar_11;

public delegate E CreateEntity<E>(string line);

public abstract class InFileRepository<ID, E> : InMemoryRepository<ID, E> where E : Entity<ID>
{
    protected string fileName;
    protected CreateEntity<E> createEntity;

    public InFileRepository(IValidator<E> vali, String fileName, CreateEntity<E> createEntity) : base(vali)
    {
        this.fileName = fileName;
        this.createEntity = createEntity;
        if (createEntity != null)
            loadFromFile();
    }

    protected virtual void loadFromFile()
    {
        List<E> list = new List<E>();
        using (StreamReader sr = new StreamReader(fileName))
        {
            string s;
            while ((s = sr.ReadLine()) != null)
            {
                E entity = createEntity(s);
                list.Add(entity);
            }
        }

        foreach (var x in list)
            entities[x.ID] = x;
    }
}